File: packages/serializer/src/bytes.ts
/**
 * Byte-level primitives of PROTOCOL.md §1: varints, zigzag, UTF-8 strings,
 * little-endian IEEE 754 floats. Used by the `schema` codec.
 *
 * Browser-safe: no Bun/Node APIs (client-js uses this too).
 */

const MAX_UINT32 = 0xffffffff

const utf8Encoder = new TextEncoder()
// ignoreBOM keeps a leading U+FEFF as data; fatal rejects invalid UTF-8.
const utf8Decoder = new TextDecoder("utf-8", { fatal: true, ignoreBOM: true })

// biome-ignore lint/suspicious/noControlCharactersInRegex: U+0000 is matched on purpose (PROTOCOL.md §1.3)
const MAY_NEED_REPLACING = /[\u0000\ud800-\udfff]/
const NOT_SCALAR =
  // biome-ignore lint/suspicious/noControlCharactersInRegex: as above
  /\u0000|[\ud800-\udbff](?![\udc00-\udfff])|(?<![\ud800-\udbff])[\udc00-\udfff]/g

/**
 * `value` as the protocol's strings hold it (PROTOCOL.md §1.3): U+0000 and
 * lone surrogates become U+FFFD, so every client, whatever its engine's
 * strings can hold, decodes the same value.
 */
export function wireString(value: string): string {
  return MAY_NEED_REPLACING.test(value)
    ? value.replace(NOT_SCALAR, "\ufffd")
    : value
}

/** Zigzag-maps an int32 to a uint32 (PROTOCOL.md §1.2). */
export function zigzag(value: number): number {
  return ((value << 1) ^ (value >> 31)) >>> 0
}

/** Inverse of {@link zigzag}. */
export function unzigzag(value: number): number {
  return (value >>> 1) ^ -(value & 1)
}

/** Strict UTF-8 decode; `undefined` for invalid input. */
export function decodeUtf8(bytes: Uint8Array): string | undefined {
  // Short ASCII strings (names, keys, ids) skip the TextDecoder call.
  if (bytes.length <= 32) {
    let ascii = ""
    for (let i = 0; i < bytes.length; i++) {
      const byte = bytes[i] ?? 0
      if (byte >= 0x80) return decodeUtf8Slow(bytes)
      ascii += String.fromCharCode(byte)
    }
    return ascii
  }
  return decodeUtf8Slow(bytes)
}

function decodeUtf8Slow(bytes: Uint8Array): string | undefined {
  try {
    return utf8Decoder.decode(bytes)
  } catch {
    return undefined
  }
}

/** A growable output buffer. Reused across encodes by its owner. */
export class ByteWriter {
  private _bytes = new Uint8Array(256)
  private _view = new DataView(this._bytes.buffer)
  private _length = 0

  public get length(): number {
    return this._length
  }

  /** Discards the contents (keeps the allocation). */
  public reset(): void {
    this._length = 0
  }

  /** Cuts the contents back to `length` bytes. */
  public truncate(length: number): void {
    this._length = length
  }

  /** A copy of the contents, owned by the caller. */
  public toBytes(): Uint8Array<ArrayBuffer> {
    return this._bytes.slice(0, this._length)
  }

  /** Overwrites one already-written byte. */
  public patch(at: number, byte: number): void {
    this._bytes[at] = byte
  }

  public u8(byte: number): void {
    this._reserve(1)
    this._bytes[this._length++] = byte
  }

  /** Unsigned LEB128; `value` must be an integer in `0..0xFFFFFFFF`. */
  public varint(value: number): void {
    this._reserve(5)
    let rest = value
    while (rest >= 0x80) {
      this._bytes[this._length++] = (rest % 0x80) | 0x80
      rest = Math.floor(rest / 0x80)
    }
    this._bytes[this._length++] = rest
  }

  /** Zigzag varint of an int32. */
  public zigzag(value: number): void {
    this.varint(zigzag(value))
  }

  public float64(value: number): void {
    this._reserve(8)
    if (Number.isNaN(value)) {
      // The canonical quiet NaN, whatever payload the engine carries.
      this._view.setUint32(this._length, 0, true)
      this._view.setUint32(this._length + 4, 0x7ff80000, true)
    } else {
      this._view.setFloat64(this._length, value, true)
    }
    this._length += 8
  }

  public float32(value: number): void {
    this._reserve(4)
    if (Number.isNaN(value)) {
      this._view.setUint32(this._length, 0x7fc00000, true)
    } else {
      this._view.setFloat32(this._length, value, true)
    }
    this._length += 4
  }

  /**
   * Varint byte length, then UTF-8. U+0000 and lone surrogates become
   * U+FFFD (PROTOCOL.md §1.3).
   */
  public string(text: string): void {
    const value = wireString(text)
    // Worst case 3 bytes per UTF-16 unit, plus the length prefix.
    this._reserve(5 + value.length * 3)
    const start = this._length
    const lengthSize = value.length < 43 ? 1 : 5
    const { written } = utf8Encoder.encodeInto(
      value,
      this._bytes.subarray(start + lengthSize),
    )
    if (lengthSize === 1) {
      this._bytes[start] = written
      this._length = start + 1 + written
      return
    }
    // Long string: write the real length, then move the bytes down to it.
    const body = start + lengthSize
    this._length = start
    this.varint(written)
    this._bytes.copyWithin(this._length, body, body + written)
    this._length += written
  }

  private _reserve(extra: number): void {
    const needed = this._length + extra
    if (needed <= this._bytes.length) return
    let size = this._bytes.length * 2
    while (size < needed) size *= 2
    const grown = new Uint8Array(size)
    grown.set(this._bytes.subarray(0, this._length))
    this._bytes = grown
    this._view = new DataView(grown.buffer)
  }
}

/**
 * Reads PROTOCOL.md §1 primitives from a body. Errors are sticky: the first
 * failed read records a message in `error`, and every later read returns a
 * zero value, so a caller checks `error` once per op instead of per read.
 */
export class ByteReader {
  public error: string | undefined
  private readonly _bytes: Uint8Array
  private readonly _view: DataView
  private _pos = 0

  public constructor(bytes: Uint8Array) {
    this._bytes = bytes
    this._view = new DataView(bytes.buffer, bytes.byteOffset, bytes.byteLength)
  }

  public get pos(): number {
    return this._pos
  }

  /** Bytes left to read. */
  public get remaining(): number {
    return this._bytes.length - this._pos
  }

  public get done(): boolean {
    return this._pos >= this._bytes.length
  }

  /** Records a failure (the first one wins). */
  public fail(message: string): void {
    if (this.error === undefined) {
      this.error = `${message} (at byte ${this._pos})`
    }
    this._pos = this._bytes.length
  }

  public u8(): number {
    const byte = this._bytes[this._pos]
    if (byte === undefined) {
      this.fail("unexpected end of data")
      return 0
    }
    this._pos++
    return byte
  }

  /** Unsigned LEB128, at most 5 bytes, at most `0xFFFFFFFF`. */
  public varint(): number {
    let value = 0
    let scale = 1
    for (let i = 0; i < 5; i++) {
      const byte = this._bytes[this._pos]
      if (byte === undefined) {
        this.fail("truncated varint")
        return 0
      }
      this._pos++
      value += (byte & 0x7f) * scale
      if ((byte & 0x80) === 0) {
        if (value > MAX_UINT32) {
          this.fail("varint above 2^32-1")
          return 0
        }
        return value
      }
      scale *= 0x80
    }
    this.fail("varint longer than 5 bytes")
    return 0
  }

  /** A varint count or length, which can't exceed the bytes left. */
  public count(): number {
    const count = this.varint()
    if (count > this.remaining) {
      this.fail(`count ${count} exceeds the ${this.remaining} bytes left`)
      return 0
    }
    return count
  }

  public zigzag(): number {
    return unzigzag(this.varint())
  }

  public float64(): number {
    if (this.remaining < 8) {
      this.fail("truncated float64")
      return 0
    }
    const value = this._view.getFloat64(this._pos, true)
    this._pos += 8
    return value
  }

  public float32(): number {
    if (this.remaining < 4) {
      this.fail("truncated float32")
      return 0
    }
    const value = this._view.getFloat32(this._pos, true)
    this._pos += 4
    return value
  }

  public string(): string {
    const length = this.count()
    if (this.error !== undefined) return ""
    const text = decodeUtf8(this._bytes.subarray(this._pos, this._pos + length))
    if (text === undefined) {
      this.fail("invalid UTF-8")
      return ""
    }
    this._pos += length
    return text
  }
}

File: clients/csharp/Bungohan.Protocol/Bytes.cs
using System;

namespace Bungohan.Protocol
{
    /// <summary>Varints (PROTOCOL.md §1.1) and zigzag (§1.2).</summary>
    public static class Varint
    {
        public const uint Max = uint.MaxValue;

        /// <summary>Bytes <paramref name="value"/> takes as a varint.</summary>
        public static int Size(uint value)
        {
            int size = 1;
            while (value >= 0x80)
            {
                value >>= 7;
                size++;
            }
            return size;
        }

        /// <summary>
        /// Reads a varint at <paramref name="offset"/>. False if it is
        /// truncated, longer than 5 bytes or above 2^32 − 1.
        /// </summary>
        public static bool TryRead(byte[] data, int offset, int end, out uint value, out int next)
        {
            ulong result = 0;
            for (int i = 0; i < 5; i++)
            {
                int at = offset + i;
                if (at >= end) break;
                byte b = data[at];
                result |= (ulong)(b & 0x7f) << (7 * i);
                if ((b & 0x80) == 0)
                {
                    next = at + 1;
                    value = (uint)result;
                    return result <= Max;
                }
            }
            value = 0;
            next = offset;
            return false;
        }

        /// <summary>Maps an int32 to a uint32: 0, −1, 1, −2 → 0, 1, 2, 3.</summary>
        public static uint Zigzag(int value) => (uint)((value << 1) ^ (value >> 31));

        /// <summary>Inverse of <see cref="Zigzag"/>.</summary>
        public static int Unzigzag(uint value) => (int)(value >> 1) ^ -(int)(value & 1);
    }

    /// <summary>
    /// A growable output buffer for the PROTOCOL.md §1 primitives. Floats
    /// are little-endian whatever the platform, and NaN is written in its
    /// canonical form.
    /// </summary>
    public sealed class ByteWriter
    {
        private byte[] _bytes;

        public ByteWriter(int capacity = 256)
        {
            _bytes = new byte[Math.Max(16, capacity)];
        }

        public int Length { get; private set; }

        public void Reset() => Length = 0;

        /// <summary>A copy of the contents.</summary>
        public byte[] ToArray()
        {
            var copy = new byte[Length];
            Buffer.BlockCopy(_bytes, 0, copy, 0, Length);
            return copy;
        }

        /// <summary>Overwrites an already-written byte.</summary>
        public void Patch(int at, byte value) => _bytes[at] = value;

        public void U8(byte value)
        {
            Reserve(1);
            _bytes[Length++] = value;
        }

        public void Bytes(byte[] data, int offset, int count)
        {
            Reserve(count);
            Buffer.BlockCopy(data, offset, _bytes, Length, count);
            Length += count;
        }

        public void Varint(uint value)
        {
            Reserve(5);
            while (value >= 0x80)
            {
                _bytes[Length++] = (byte)(value | 0x80);
                value >>= 7;
            }
            _bytes[Length++] = (byte)value;
        }

        public void Zigzag(int value) => Varint(Protocol.Varint.Zigzag(value));

        public void Float64(double value)
        {
            long bits = double.IsNaN(value) ? 0x7ff8000000000000L : BitConverter.DoubleToInt64Bits(value);
            Reserve(8);
            for (int i = 0; i < 8; i++) _bytes[Length++] = (byte)(bits >> (8 * i));
        }

        /// <summary>Writes <paramref name="value"/> rounded to binary32 (ties to even).</summary>
        public void Float32(double value)
        {
            int bits = double.IsNaN(value) ? 0x7fc00000 : BitConverter.SingleToInt32Bits((float)value);
            Reserve(4);
            for (int i = 0; i < 4; i++) _bytes[Length++] = (byte)(bits >> (8 * i));
        }

        /// <summary>Varint byte length, then the UTF-8 bytes.</summary>
        public void String(string value)
        {
            byte[] utf8 = Utf8.Encode(value);
            Varint((uint)utf8.Length);
            Bytes(utf8, 0, utf8.Length);
        }

        /// <summary>Big-endian unsigned integer of <paramref name="size"/> bytes (MessagePack).</summary>
        public void BigEndian(ulong value, int size)
        {
            Reserve(size);
            for (int i = size - 1; i >= 0; i--) _bytes[Length++] = (byte)(value >> (8 * i));
        }

        private void Reserve(int extra)
        {
            int needed = Length + extra;
            if (needed <= _bytes.Length) return;
            int size = _bytes.Length * 2;
            while (size < needed) size *= 2;
            Array.Resize(ref _bytes, size);
        }
    }

    /// <summary>
    /// Reads PROTOCOL.md §1 primitives. Errors are sticky: the first failed
    /// read records <see cref="Error"/>, and every later read returns a zero
    /// value, so callers check once per op rather than per read.
    /// </summary>
    public sealed class ByteReader
    {
        private readonly byte[] _bytes;
        private readonly int _end;

        public ByteReader(byte[] bytes) : this(bytes, 0, bytes.Length) { }

        public ByteReader(byte[] bytes, int offset, int count)
        {
            _bytes = bytes;
            Position = offset;
            _end = offset + count;
        }

        public string? Error { get; private set; }
        public int Position { get; private set; }
        public int Remaining => _end - Position;
        public bool Done => Position >= _end;
        public bool Failed => Error != null;

        /// <summary>Records a failure (the first one wins) and stops reading.</summary>
        public void Fail(string message)
        {
            if (Error == null) Error = message + " (at byte " + Position + ")";
            Position = _end;
        }

        public byte U8()
        {
            if (Position >= _end)
            {
                Fail("unexpected end of data");
                return 0;
            }
            return _bytes[Position++];
        }

        public uint Varint()
        {
            if (Protocol.Varint.TryRead(_bytes, Position, _end, out uint value, out int next))
            {
                Position = next;
                return value;
            }
            Fail("bad varint (truncated, longer than 5 bytes, or above 2^32-1)");
            return 0;
        }

        /// <summary>A count or length, which can't exceed the bytes left.</summary>
        public int Count()
        {
            uint count = Varint();
            if (count > (uint)Remaining)
            {
                Fail("count " + count + " exceeds the " + Remaining + " bytes left");
                return 0;
            }
            return (int)count;
        }

        public int Zigzag() => Protocol.Varint.Unzigzag(Varint());

        public double Float64()
        {
            if (Remaining < 8)
            {
                Fail("truncated float64");
                return 0;
            }
            long bits = 0;
            for (int i = 0; i < 8; i++) bits |= (long)_bytes[Position + i] << (8 * i);
            Position += 8;
            return BitConverter.Int64BitsToDouble(bits);
        }

        public double Float32()
        {
            if (Remaining < 4)
            {
                Fail("truncated float32");
                return 0;
            }
            int bits = 0;
            for (int i = 0; i < 4; i++) bits |= _bytes[Position + i] << (8 * i);
            Position += 4;
            return BitConverter.Int32BitsToSingle(bits);
        }

        public string String()
        {
            int length = Count();
            if (Failed) return "";
            string? text = Utf8.Decode(_bytes, Position, length);
            if (text == null)
            {
                Fail("invalid UTF-8");
                return "";
            }
            Position += length;
            return text;
        }

        /// <summary>Big-endian unsigned integer of <paramref name="size"/> bytes (MessagePack).</summary>
        public ulong BigEndian(int size)
        {
            if (Remaining < size)
            {
                Fail("truncated value");
                return 0;
            }
            ulong value = 0;
            for (int i = 0; i < size; i++) value = (value << 8) | _bytes[Position + i];
            Position += size;
            return value;
        }

        /// <summary>A copy of the next <paramref name="count"/> bytes.</summary>
        public byte[] Take(int count)
        {
            if (Remaining < count)
            {
                Fail("truncated value");
                return Array.Empty<byte>();
            }
            var copy = new byte[count];
            Buffer.BlockCopy(_bytes, Position, copy, 0, count);
            Position += count;
            return copy;
        }

        /// <summary>Decodes UTF-8 of <paramref name="count"/> bytes (MessagePack strings).</summary>
        public string Utf8String(int count)
        {
            if (Remaining < count)
            {
                Fail("truncated string");
                return "";
            }
            string? text = Utf8.Decode(_bytes, Position, count);
            if (text == null)
            {
                Fail("invalid UTF-8");
                return "";
            }
            Position += count;
            return text;
        }
    }
}
